[indvars] Missing variables at Og

CarlosAlbertoEnciso opened the pull request:
https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

In the given test case:
 ```
4 ...
 5 void bar() {
 6   int End = 777;
 7   int Index = 27;
 8   char Var = 1;
 9   for (; Index < End; ++Index)
10     ;
11   nop(Index);
12 }
13 ...
```
Missing local variable `Index` after loop `Induction Variable Elimination`.
When adding a breakpoint at line `11`, LLDB does not have information
on the variable. But it has info on `Var` and `End`.

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

In the given test case:

 4 ...
 5 void bar() {
 6   int End = 777;
 7   int Index = 27;
 8   char Var = 1;
 9   for (; Index < End; ++Index)
10     ;
11   nop(Index);
12 }
13 ...

Missing local variable 'Index' after loop 'Induction Variable Elimination'.
When adding a breakpoint at line 11, LLDB does not have information on
the variable. But it has info on 'Var' and 'End'.

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

In the given test case:

 4 ...
 5 void bar() {
 6   int End = 777;
 7   int Index = 27;
 8   char Var = 1;
 9   for (; Index < End; ++Index)
10     ;
11   nop(Index);
12 }
13 ...

Missing local variable 'Index' after loop 'Induction Variable Elimination'.
When adding a breakpoint at line 11, LLDB does not have information on
the variable. But it has info on 'Var' and 'End'.

- Moved the new logic to 'rewriteLoopExitValues'.
- Removed from the test cases, the unrelated functions: 'nop' and 'main'.
- Use 'isa<SCEVConstant>'.

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

In the given test case:

 4 ...
 5 void bar() {
 6   int End = 777;
 7   int Index = 27;
 8   char Var = 1;
 9   for (; Index < End; ++Index)
10     ;
11   nop();
12 }
13 ...

Missing local variable 'Index' after loop 'Induction Variable Elimination'.
When adding a breakpoint at line 11, LLDB does not have information on
the variable. But it has info on 'Var' and 'End'.

Address reviewers comments.
- Early exit to simplify the logic.
- Avoid inserting the same instruction in multiple blocks.
- Skip debug-users with variadic variable locations.
- Change some comments to improve readability.
- Add code to clone and move the debug value.
- Modify second test case to include multiple exit blocks.

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

In the given test case:

 4 ...
 5 void bar() {
 6   int End = 777;
 7   int Index = 27;
 8   char Var = 1;
 9   for (; Index < End; ++Index)
10     ;
11   nop();
12 }
13 ...

Missing local variable 'Index' after loop 'Induction Variable Elimination'.
When adding a breakpoint at line 11, LLDB does not have information on
the variable. But it has info on 'Var' and 'End'.

Address reviewers comments.
- Early exit to simplify the logic.
- Avoid inserting the same instruction in multiple blocks.
- Skip debug-users with variadic variable locations.
- Change some comments to improve readability.
- Add code to clone and move the debug value.
- Modify second test case to include multiple exit blocks.

Addressed the upstream feedback in relation to:
- Each exit block has its own exit value.
- Separate the debug values for incoming and exit values.

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

Fix clang-format found issues in updated patch.

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

Addressed the upstream feedback in relation to:
- Remove the introduced 'ExitBlock' field.
- Update some comments.

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

After internal discussion with @jmorse, it was decided
to split the work between the 'indvars' and the
'loop-deletion' passes.

1) passes="loop(indvars)"
- 'indvars' transformation is fired:
  the 'rewriteLoopExitValues' will rewrite the collected
  PNs with the exit values.
- 'indvars' transformation is not fired:
  If the loop can be deleted, we preserve the induction
  variable information to be used by 'loop-deletion' if
  that pass will be executed.

2) passes="loop(indvars,loop-deletion)"
  If the loop is deleted in 'deleteDeadLoop' and there
  is a valid exit block, we use any collected values
  by 'indvars' to updated the exit values.

Added extra tests to cover the following cases:
  ...
  char Var = 1;
  for (; Index < End; ++Index)
    if (Index == 666)
      ++Var;
  ...

and
  ...
  char Var = 1;
  for (; Index < End; ++Index)
    if (Index == 666)
      Var = 555;
  ...

Modified but otherwise unused variable 'Var' in a loop
that gets deleted.

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

Address @SLTozer comments.
- Use captured variables for SSA values.
- Use new insert logic calls.
- Remove not required 'CHECK-NOT'.

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

Address @SLTozer comments.
- Don't use any names for not used SSA values.

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

Address @nikic comments about layering violation.
- Move code to 'LoopInfo.h'.

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

Fix Clang-format issue.

OCHyams commented on llvm/include/llvm/Analysis/LoopInfo.h:
```diff
@@ -392,13 +394,35 @@ class LLVM_EXTERNAL_VISIBILITY Loop : public LoopBase<BasicBlock, Loop> {
+  /// 'indvars' pass if the loop can deleted. Those debug users will be used
+  /// by the 'loop-delete' pass.
+  void preserveDebugInductionVariableInfo(
+      Value *FinalValue, SmallVector<DbgVariableIntrinsic *> DbgUsers) {
```
I think it makes sense to delete the DbgUsers parameter and do the `findDbgUsers` call from within this function?

(If you disagree, please update `DbgUsers` to be a const SmallVectorImpl reference or ArrayRef)

CarlosAlbertoEnciso replied:
To avoid adding a dependency on `DebugInfo.h`, making `DbgUsers` to be `const SmallVectorImpl reference`.

OCHyams commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -1611,11 +1657,30 @@ int llvm::rewriteLoopExitValues(Loop *L, LoopInfo *LI, TargetLibraryInfo *TLI,
+        if (L->getUniqueExitBlock()) {
+          SmallVector<DbgVariableIntrinsic *> DbgUsers;
+          findDbgUsers(DbgUsers, IndVar);
+          L->preserveDebugInductionVariableInfo(FinalIVValue, DbgUsers);
```
Is there a reason that we can't insert a dbg.value in the exit block right now, here, rather than waiting for loop-deletion to delete the loop?

nikic replied:
TBH I still think the whole design here makes little sense. I think this case should either just be Won't Fix, or LoopDeletion needs to compute the exit value, rather than relying on some complex interaction with IndVars (which is not the only user of LoopDeletion!)

CarlosAlbertoEnciso replied:
@nikic Patch already reverted.

If I understand correctly, your main concern is the interaction between `IndVars` and `LoopDeletion` passes.

What I would suggest, is to divide the patch into 2 independent ones:
1) If the `IndVars` pass is triggered add the missing debug values.
2) `LoopDeletion` to compute the exit value.

With no interaction between those 2 passes.

Your feedback would be highly valuable.

OCHyams reviewed (commented):
I've not reviewed this in depth yet - just left some first-pass comments (one surface level, one more important asking about why this needs to be a multi-pass fix).

CarlosAlbertoEnciso commented:
> I've not reviewed this in depth yet - just left some first-pass comments (one surface level, one more important asking about why this needs to be a multi-pass fix).

The main discussion is here:
https://github.com/llvm/llvm-project/pull/69920#issuecomment-1905988593

If you prefer, I can elaborate more if is not clear.

OCHyams commented:
> The main discussion is here:
> https://github.com/llvm/llvm-project/pull/69920#issuecomment-1905988593

Thanks, I'd missed that this is a re-posting of #69920, which already had approval. My other small inline comment still stands though (the "surface level" one). And I think this needs clang-formatting? Otherwise SGTM.

CarlosAlbertoEnciso committed:
[indvars] Missing variables at Og:

https://bugs.llvm.org/show_bug.cgi?id=51735
https://github.com/llvm/llvm-project/issues/51077

Address @OCHyams comments about DbgUsers type.
- Converted to 'const SmallVectorImpl reference'.

CarlosAlbertoEnciso commented:
@OCHyams Thanks for your review.

nikic commented on llvm/include/llvm/Analysis/LoopInfo.h:
```diff
@@ -17,7 +17,9 @@
 #include "llvm/IR/Instructions.h"
+#include "llvm/IR/IntrinsicInst.h"
 #include "llvm/IR/PassManager.h"
+#include "llvm/IR/ValueHandle.h"
```
Adding these includes to a core header like LoopInfo.h adds significant compile-time overhead.

CarlosAlbertoEnciso commented:
@nikic Thanks for your valuable feedback. 
I will revert the patch.
